src/components/document-viewer/DocFragment.jsx

```diff
@@ -22,26 +25,86 @@ class DocFragment extends Component {
   constructor(props) {
     super(props);
 
+    const json = {};
+    json[this.props.title] = _.pickBy(
+      _.mapValues(this.props.content, x => (_.isDate(x) ? x.toString() : x)),
+      _.negate(_.isFunction),
+    );
+    const sourceCode = jsyaml.safeDump(json);
+
     this.state = {
       showEditor: false,
+      lintError: null,
+      validationError: null,
+      sourceCode,
+      message: '',
     };
 
     this.saveChange = this.saveChange.bind(this);
   }
 
-  saveChange() {
-    const source = this.refs.ace.editor.getValue();
-    DocumentService.updateFragment(this.props.content.getHash(), source);
+  cancelEdit() {
+    const json = {};
+    json[this.props.title] = _.pickBy(
+      _.mapValues(this.props.content, x => (_.isDate(x) ? x.toString() : x)),
+      _.negate(_.isFunction),
+    );
+    const sourceCode = jsyaml.safeDump(json);
 
     this.setState({
-      loading: true,
+      showEditor: false,
+      sourceCode,
     });
-    GitHubService.commit({ content: DocumentService.getSource() }).then(() =>
-      this.setState({
-        loading: false,
-        showEditor: false,
-      }),
-    );
+  }
+
+  handleChange() {
+    const sourceCode = this.refs.ace.editor.getValue();
+    this.setState({ sourceCode });
+    lint(sourceCode)
+      .then(() => {
+        this.setState({ lintError: null });
+      })
+      .catch((err) => {
+        this.setState({ lintError: err.message });
+      });
+  }
+
+  updateCommitMessage(e) {
+    const message = e.target.value;
+    this.setState({ message });
+  }
+
+  saveChange() {
+    const source = this.refs.ace.editor.getValue();
+    const message = `Edited section "${this.props.title}" using Pensieve\n\n${this.state.message}`;
+    const { schema } = store.getState();
+
+    lint(source)
+      // When validating, strip the title field from the source
+      .then(() => schemaValidate(schema, source.replace(/^.+\n/, '')))
+      .then(() => {
+        DocumentService.updateFragment(this.props.content.getHash(), source);
+
+        this.setState({
```

Comment:
Not sure what the output of loading is but is it not better to set it right away? Before `lint`?